src/mqt/predictor/reward.py

```diff
@@ -69,3 +70,90 @@ def calc_qubit_index(qargs: list[Qubit], qregs: list[QuantumRegister], index: in
+                    pass  # gate duration not available for this qubit pair
+        else:  # single-qubit gates
+            for q in range(device.num_qubits):
+                try:
+                    if op == "measure":
+                        d = device.get_readout_duration(q)
+                    else:
+                        d = device.get_single_qubit_gate_duration(op, q)
+                    op_times.append((op, [q], d, "s"))
+                except ValueError:  # noqa:PERF203
+                    pass  # gate duration not available for this qubit
+
+    # associate gate times for each qubit through asap scheduling
+    transpiled = transpile(qc, scheduling_method="asap", basis_gates=device.basis_gates, instruction_durations=op_times)
+
+    qubit_durations: dict[int, list[float]] = {qubit_index: [] for qubit_index in range(device.num_qubits)}
+
+    for instruction, qargs, _cargs in transpiled.data:
+        gate_type = instruction.name
+        if gate_type == "barrier":
+            continue
+
+        assert len(qargs) in [1, 2]
+        first_qubit_idx = calc_qubit_index(qargs, qc.qregs, 0)
+        if device.name == "rigetti_aspen_m2":
+            first_qubit_idx = (first_qubit_idx + 40) % 80
+
+        fidelity = 1.0
+        if len(qargs) == 1:  # single-qubit gate
+            if gate_type == "measure":  # measurement reliability
+                fidelity = device.get_readout_fidelity(first_qubit_idx)
+            elif gate_type == "delay":
+                pass  # idle: only T1,T2 decoherence
+            else:  # gate reliability
+                fidelity = device.get_single_qubit_gate_fidelity(gate_type, first_qubit_idx)
+            qubit_durations[first_qubit_idx].append(instruction.duration)
+
+        else:  # multi-qubit gate
+            second_qubit_idx = calc_qubit_index(qargs, qc.qregs, 1)
+            if device.name == "rigetti_aspen_m2":
+                second_qubit_idx = (second_qubit_idx + 40) % 80
+
+            fidelity = device.get_two_qubit_gate_fidelity(gate_type, first_qubit_idx, second_qubit_idx)
+            qubit_durations[first_qubit_idx].append(instruction.duration)
+            qubit_durations[second_qubit_idx].append(instruction.duration)
+
+        res *= fidelity
+
+    # calculate T1,T2 decoherence over entire qubit lifetime
+    for qubit_idx, durations in qubit_durations.items():
+        # don't consider idle qubits
+        if len(durations) <= 1:
+            continue
+        lifetime = sum(durations)  # all operation times on qubit
+        # only use dominant decoherence term (either T1 or T2) as in https://arxiv.org/pdf/2001.02826
+        # alternative: np.exp(-lifetime / T_1 - lifetime / T_2) as in https://arxiv.org/abs/2306.15020
+        T_min = min(device.calibration.get_t1(qubit_idx), device.calibration.get_t2(qubit_idx))  # noqa:N806
+        decoherence = np.exp(-lifetime / T_min)
+        res *= decoherence
+
+    return cast(float, np.round(res, precision))
```

Comment:
I have not fully understood this method, could you give me a brief introduction in person?